projects/js-packages/publicize-components/src/components/share-status/use-refresh-post-share-status.ts

```diff
@@ -0,0 +1,77 @@
+import { useDispatch, useSelect } from '@wordpress/data';
+import { useCallback, useEffect, useState } from '@wordpress/element';
+import { store as socialStore } from '../../social-store';
+import { ONE_MINUTE_IN_MS } from './constants';
+
+export type RefreshPostShareStatusOptions = {
+	postId?: number;
+	timeout?: number;
+};
+
+/**
+ * Refresh post share status.
+ *
+ * @param {RefreshPostShareStatusOptions} options - Options.
+ *
+ * @return {() => void} - Function to start polling.
+ */
+export function useRefreshPostShareStatus( {
+	postId,
+	timeout = ONE_MINUTE_IN_MS,
+}: RefreshPostShareStatusOptions = {} ) {
+	// @ts-expect-error `invalidateResolution` exists in every store
+	const { invalidateResolution } = useDispatch( socialStore );
+
+	const postShareStatus = useSelect(
+		select => select( socialStore ).getPostShareStatus( postId ),
+		[]
+	);
+
+	const [ hasTimeoutPassed, setHasTimeoutPassed ] = useState( false );
+	const [ pollForShareStatus, setPollForShareStatus ] = useState( false );
+	const [ lastTimestamp, setLastTimestamp ] = useState( 0 );
+
+	// If the last timestamp is present, check if there are newer timestamps.
+	// If the last timestamp is not present, check if we have any shares.
+	const isRequestComplete = lastTimestamp
+		? postShareStatus.shares.some( share => share.timestamp > lastTimestamp )
+		: postShareStatus.shares.length > 0;
+
+	useEffect( () => {
+		// Stop polling if the request is complete.
+		if ( isRequestComplete ) {
+			setPollForShareStatus( false );
+			setLastTimestamp( 0 );
+			return;
+		}
+
+		if ( ! pollForShareStatus || hasTimeoutPassed || postShareStatus.loading ) {
+			return;
+		}
+		invalidateResolution( 'getPostShareStatus', [ postId ] );
+	}, [
+		pollForShareStatus,
+		isRequestComplete,
+		hasTimeoutPassed,
+		postShareStatus.loading,
+		invalidateResolution,
+		postId,
+	] );
+
+	return useCallback( () => {
+		// Fist, save the last timestamp.
+		setLastTimestamp(
+			// Since the shares are arranged in descending order of timestamp, the first share has the latest timestamp.
+			postShareStatus.shares[ 0 ]?.timestamp || 0
+		);
+
+		// Then invalidate the resolution to start fetching.
```

Comment:
Is this the bit that controls the polling? If yes, why we don't we wrap it in a settimeout? That way we don't we call the endpoint so many times.

Reply:
Yes, there are different ways to achieve what we want but I think it needs to be handled in a different PR for the post publish panel as well which uses the same approach.

Reply:
Shall we take the time and figure that out in a different PR? We can work on that solution and when we are happy with it, we can have both of these merged in quick succession. 

This PR seems risky to be merged in its current format. I am aware there is a blog sticker, but I don't think it's worth the risk.

Reply:
Should we then revert the change we made for post publish panel that does the same?

We can then look for a better solution for both the places.

Reply:
No, let's not revert things if it's already in there, we are past the code freeze anyway.